### PR TITLE
Adds support for "typesVersions" redirects

### LINKED ISSUE #22605: Provide a back-compat mechanism for bundled .d.ts files in package.json
People are always excited to use new type system features in their programs or definition files.

However, if these developers are shipping a library with bundled type definitions, they can't use these features in an exposed way unless they are willing to force all TS consumers of the library to upgrade to a matching compiler version.

This leads to awkward trade-offs and end-user confusion when they upgrade eventually does happen.

There should be some way for library authors to say "Use *this* .d.ts file if you're on *this* version of TypeScript" so they can provide back-compat ranges. This could be based on semver ranges provided in `package.json`:

```
{
  ...
  "types-compat": {
    "<=2.3": "backcompat/index-2.3.d.ts",
    "<=2.7": "backcompat/index-2.7.d.ts",
    "*": "index.d.ts"
  }
}
```

### PR DESCRIPTION
**Pull request by rbuckton**
This adds support TypeScript version-specific redirects to type definitions in a package. If a package wishes to include type definitions that gracefully downgrade to older editions of TypeScript, they will be able to do so by adding the following entry to their _package.json_ file:

```json
{
  "name": "package-name",
  "version": "1.0",
  "types": "index",
  "typesVersions": {
    ">=3.1.0-0": { "*": ["ts3.1/*"] }
  }
}
```

When a `"typesVersions"` field is found in a referenced package, we search for the first version range that matches the current compiler version. If a match is found, the associated path patterns are applied.

Version ranges specified in `"typesVersions"` keys follow the same range comparator format (strict only) as https://github.com/npm/node-semver#ranges.

Path patterns are applied in the same way that the `"paths"` compiler option is applied to paths in your project.

This feature will only support compiler version of 3.1 or later. Earlier TypeScript compilers will use the existing behavior.

Fixes #22605

### CODE REVIEW

**Comment by rbuckton** [comment]
There is some overlap between _src/compiler/semver.ts_ and _src/jsTyping/semver.ts_. I may retire the version in jsTyping as its very specific to its implementation, while the implementation in compiler is more generic and can serve both cases. I can either address this in this PR or a follow up.

**Comment by weswigham** [comment]
> If a match is found, the value is inserted into the path between the package directory and "types" (or any named submodule beneath the package).

This feels strange. I'd expect either a module specifier (so it points directly at another package to defer to), or a direct path.

**Comment by rbuckton** [comment]
This is more like a per-version root directory override. There are two kinds of typed packages we need to support:

- Ambient module based packages (that generally have types defined in a single file like "index.d.ts")
- Non-ambient module based packages (that use top level exports and use multiple files)

See https://gist.github.com/rbuckton/bcde1c2adde270c3b0d7ced5320c0877 for a more detailed description and examples.

**Comment by rbuckton** [comment]
> […] I'd expect either a module specifier (so it points directly at another package to defer to), or a direct path.

Pointing to another package is not the intended scenario. This is intended to support the following two scenarios:

1. Declarations defined via ambient modules in a single _index.d.ts_ file (common on DefinitelyTyped)
2. Declarations defined via one or more declaration files using top level exports (common on DefinitelyTyped and default behavior of TypeScript build outputs).

If it were a direct path, we could not support (2). We could consider some kind of path mapping/path template, but overall it would just add unnecessary complexity.

**Comment by rbuckton** [comment]
I have a few open questions:

- Should we ignore paths that traverse outside of the package (e.g. disallow absolute paths or "..")?
- Should we compare against TypeScript's `versionMajorMinor` or `version`? While `version` seems to be more correct, due to the precedence rules of semver, `3.1` comes _after_ `3.1.0-dev`/`3.1.0-next`, so using a simple version number like `3.1` wouldn't work when running a dev/next/insiders build.
- Should we consider a path templating feature, or leave it as a prefix path only?

**Comment by rbuckton** [comment]
If we decide we want to support semver ranges in `"typesVersions"`, I actually was able to put together a relatively lightweight semver range parser based on the BNF here: https://github.com/npm/node-semver#range-grammar

I put a copy of the version of semver.ts with support for ranges here [in a gist]( https://gist.github.com/rbuckton/5b44e39ce6bb095d3df452e21a5ac653) so as not to pollute this PR with this change unless we decide to implement it.

**Comment by rbuckton** [comment]
> Should we consider a path templating feature, or leave it as a prefix path only?

If we want to leverage our existing path templating functionality, it would change `"typesVersions"` to look like this:

```json
{
  "typesVersions": {
    "3.1": { "*": ["ts3.1/*"] }
  }
}
```

This would also require refactoring `tryLoadModuleUsingBaseUrl` to extract its logic and inject it in various places in _moduleNameResolver.ts_.

**Comment by weswigham** [comment]
Ah! Also a test of what we do when a package's "latest version" root file attempts to `import` the real root, eg
```ts
// @declaration: true
// @filename: node_modules/bar/tsnew/index.d.ts
export * from "../";
// @filename: node_modules/bar/tsnew/deep.d.ts
export * from "../deep";
// @filename: node_modules/bar/index.d.ts
export interface Bar {}
export function bar(): Bar;
// @filename: node_modules/bar/deep.d.ts
export interface Bar2 {}
export function bar2(): Bar2;
// @filename: node_modules/bar/package.json
{
  "name": "bar",
  "version": 0.0.1,
  "types": "index.d.ts",
  "typesVersions": { "3.1": "tsnew/" }
}
// @filename: consumer.ts
import { bar } from "bar";
import { bar2 } from "bar/deep";
export const v = bar();
export const v2 = bar2();
```

**Comment by weswigham** [comment]
Also, I know I've said this in person, but do we know of any non-DT people we're writing this for? Because on DT this is all going to be managed by the gen script, so no human hand should ever touch these fields - meaning a pretty or simple configuration isn't required; a manifest of real paths to redirects would even be sufficient, eg
```json
{
  "typesVersions": {
    "3.1": {
      "./index.d.ts": "./ts3.1/index.d.ts",
      "./deep.d.ts": "./ts3.1/deep.d.ts",
      "./sub/module/index.d.ts": "./ts3.1/sub/module/index.d.ts",
    }
  }
}
```
which avoids encoding the path-only-redirects convention into the feature without going full path-mapping. If we know of some others who'd like to use this, it'd be great to get their input on what kind of configuration makes the most sense from the perspective of their build tools and source trees.

**Comment by ajafff** [comment]
> do we know of any non-DT people we're writing this for? 

👋 
I need this in `tsutils` to expose an API that is compatible with the TypeScript version you are compiling with.

Regarding submodule imports:
When importing from `tsutils/util` does it even look at `tsutils/package.json`? Node's module spec says it doesn't.

However, I want to redirect imports from package root as well as imports from any submodule:
* `tsutils` -> `tsutils/${tsVersion}`
* `tsutils/typeguard/node` -> `tsutils/${tsVersion}/typeguard/node`

**Comment by rbuckton** [comment]
If we support path mappings, would we want completions to include both mapped and unmapped paths?

Consider the directory structure for **node_modules/foo**:
```
package.json
index.d.ts
aaa.d.ts
ts3.1/index.d.ts
ts3.1/zzz.d.ts
```

With the following in **package.json**:

```json
{
  "typesVersions": {
    "3.1": { "*": ["ts3.1/*"] }
  }
}
```

If we follow the same logic for completions that we do for baseUrl/paths, then a completion for `"foo/` would return `index`, `aaa`, `ts3.1` (directory completion), and `zzz` (from path mapping). 

Does that seem acceptable?

**Comment by rbuckton** [comment]
I still need a test for https://github.com/Microsoft/TypeScript/pull/26568#issuecomment-415237604, but I imagine it we would want it to work the same way that baseUrl/paths works for this case.

**Comment by rbuckton** [comment]
> Also a test of what we do when a package's "latest version" root file attempts to import the real root […]
>
> ```ts
> // @declaration: true
> // @filename: node_modules/bar/tsnew/index.d.ts
> export * from "../";
> ```

Due to the fact the path mapping applies to the whole package, **bar/tsnew/index.d.ts** effectively refers to itself (since **bar/tsnew/..** refers to **bar**, which gets redirected back to **bar/tsnew/index.d.ts**). The end result is a module with no exports. I think this may be acceptable since you should avoid writing your module this way anyways.

**Comment by RyanCavanaugh** [comment]
@kitsonk @nicknisi @agubler I recall Dojo having problems with bundled .d.ts files + TS compat - FYI in case you want to weigh in on this proposal

**Comment by rbuckton** [comment]
As of dc5cd96 the version keys now use the same semver range format used in other places in **package.json**. 

It is important to note that while a range of `3.1` would match public releases of TypeScript, it would not match prerelease builds (e.g. `-dev` or `-insiders` builds, etc.). The example uses the version range `">=3.1.0-0"`, as this specifies a minimum version of any version of TypeScript starting at 3.1 _including prereleases_ (due to the additional `-0` qualifier in the version). This aligns with the same behavior of semver in package.json and with npm.

**Comment by rbuckton** [comment]
@weswigham can you take another look?

File: src/compiler/moduleNameResolver.ts [hunk 1]
```diff
@@ -96,48 +105,103 @@ namespace ts {
...
-            if (result.error) {
-                return {};
+    export function getPackageJsonTypesVersionsPaths(typesVersions: MapLike<MapLike<string[]>>) {
+        if (!typeScriptVersion) typeScriptVersion = new Version(version);
+
+        for (const key in typesVersions) {
+            if (!hasProperty(typesVersions, key)) continue;
+
+            const keyRange = VersionRange.tryParse(key);
+            if (keyRange === undefined) {
```
Comment (RyanCavanaugh) [inline src/compiler/moduleNameResolver.ts]: This case probably deserves a log message since it's basically a silent error

**Comment by dball-adashi** [comment]
Backward/Forward .d.ts compatibility system #31907

**Comment by dball-adashi** [comment]
#31907

**Comment by dball-adashi** [comment]
#31907

**Pull request merged**